scalyr_agent/agent_main.py

```diff
@@ -689,12 +706,93 @@ def __handle_terminate(self):
             log.info("Received signal to shutdown, attempt to shutdown cleanly.")
             self.__run_state.stop()
 
+    def _get_system_and_agent_stats(self):  # type: () -> Optional[Dict]
+        """
+        Return current machine's and agent process' stats.
+        :return: Dict with stats.
+        """
+
+        # Get information about agent process.
+        def get_agent_process_stats():
+            agent_process_stats = {}
+            if psutil is None:
+                return {"error": "No psutil"}
+
+            pidfile = os.path.join(self.__config.agent_log_path, "agent.pid")
+
+            if not os.path.exists(pidfile):
+                return {"error": "No PID file."}
+
+            try:
+                with open(pidfile, "r") as fp:
+                    pidfile_content = fp.read().strip()
+            except OSError as e:
+                return {"error": "Error during PID file read. Error: {}".format(e)}
+
+            try:
+                psutil_process = psutil.Process(pid=int(pidfile_content))
+            except psutil.Error as e:
+                return {"error": "Can't create psutil process. Error: {}".format(e)}
+
+            try:
+                agent_process_stats["cpu_percent"] = psutil_process.pid
+                agent_process_stats["cpu_percent"] = psutil_process.cpu_percent()
+                agent_process_stats["cpu_percent"] = psutil_process.memory_info().rss
+            except Exception as e:
+                agent_process_stats["error"] = str(e)
+
+            return agent_process_stats
+
+        # Get information about machine's current state.
+        def get_machine_stats():
+            machine_stats = {}
+            try:
+                machine_stats["cpu_percent"] = psutil.cpu_percent()
+                machine_stats["cpu_count"] = psutil.cpu_count()
+                machine_stats["getloadavg"] = psutil.getloadavg()
+                machine_stats[
+                    "virtual_memory_percent"
+                ] = psutil.virtual_memory().percent
+            except Exception as e:
+                machine_stats["error"] = str(e)
+
+            return machine_stats
+
+        # Get list of related processes'
+        def find_agent_processed():
+            agent_processes = []
+            try:
+                ps_output = subprocess.check_output(
+                    [
+                        "ps",
+                        "aux",
+                    ]
+                ).decode()
+            except Exception as e:
+                return {"error": "Can't get list of processes. Error: {}".format(e)}
+
+            for line in ps_output.splitlines():
+                if "scalyr-agent" in line or "python" in line or "agent_main" in line:
+                    agent_processes.append(line)
+
+            return agent_processes
+
+        result = {}
```

Comment:
I wonder if it makes sense to also try to include the output of equivalent of ``ls -la status_file`` and ``cat status_file`` here.

In case there is some race condition with writing / reading a file, this may make it easier to catch it.